### PR TITLE
Sol Balance VC with an abort strategy example implementation

### PR DESCRIPTION
**Pull request by wli-pro**
### Labels

Please apply following PR-related labels when appropriate:
- `C0-breaking`: if your change could break the existing client, e.g. API change, critical logic change 
- `C1-noteworthy`: if your change is non-breaking, but is still worth noticing for the client, e.g. reference code improvement

### Testing Evidences

Please attach any relevant evidences if applicable

```
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://w3id.org/security/suites/ed25519-2020/v1"
  ],
  "id": "0x972899b24a5989a3ae77e5b438ba6f7c2cebe2af15ccb32900d48145eb30250b",
  "type": [
    "VerifiableCredential"
  ],
  "credentialSubject": {
    "id": "did:litentry:substrate:0xd4e35b16ec6b417386b948e7eaf5cc642a243096cecf366e6313689b90969f42",
    "description": "The amount of a particular token you are holding",
    "type": "Token Holding Amount",
    "assertionText": "TokenHoldingAmount(Sol)",
    "assertions": [
      {
        "and": [
          {
            "src": "$token",
            "op": "==",
            "dst": "SOL"
          },
          {
            "or": [
              {
                "and": [
                  {
                    "src": "$network",
                    "op": "==",
                    "dst": "bsc"
                  },
                  {
                    "src": "$address",
                    "op": "==",
                    "dst": "0x570a5d26f7765ecb712c0924e4de545b89fd43df"
                  }
                ]
              },
              {
                "and": [
                  {
                    "src": "$network",
                    "op": "==",
                    "dst": "ethereum"
                  },
                  {
                    "src": "$address",
                    "op": "==",
                    "dst": "0x5288738df1aeb0894713de903e1d0c001eeb7644"
                  }
                ]
              },
              {
                "and": [
                  {
                    "src": "$network",
                    "op": "==",
                    "dst": "solana"
                  }
                ]
              }
            ]
          },
          {
            "src": "$holding_amount",
            "op": ">=",
            "dst": "1"
          },
          {
            "src": "$holding_amount",
            "op": "<",
            "dst": "50"
          }
        ]
      }
    ],
    "values": [
      true
    ],
    "endpoint": "http://localhost:9933"
  },
  "issuer": {
    "id": "did:litentry:substrate:0xefe8ed375ee8dde2cbc6fe63b39fbf0beca657e263d35db84c2c62a970e6897e",
    "name": "Litentry TEE Worker",
    "mrenclave": "FeWRdXbXksaPWP6cpptyv9kuCHLyixpv3v7muhEFtK6A"
  },
  "issuanceDate": "2024-03-06T13:50:46.233886365+00:00",
  "parachainBlockNumber": 87,
  "sidechainBlockNumber": 163,
  "proof": {
    "created": "2024-03-06T13:50:46.235452419+00:00",
    "type": "Ed25519Signature2020",
    "proofPurpose": "assertionMethod",
    "proofValue": "c5a7155e52f066bbfaf48f1b0fc9e6c5dcc2684b84a30990c12692a84fb5036820a8fefd20aff6d341b2945fd1cff378b8d627aa0d2924e1d2b6d3c09755ab00",
    "verificationMethod": "0xefe8ed375ee8dde2cbc6fe63b39fbf0beca657e263d35db84c2c62a970e6897e"
  },
  "credentialSchema": {
    "id": "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/21-evm-holding-amount/1-0-0.json",
    "type": "JsonSchemaValidator2018"
  }
}
```

### CODE REVIEW

File: bitacross-worker/docker/docker-compose.yml [hunk 1]
```diff
@@ -130,6 +130,7 @@ services:
       - SORA_QUIZ_MASTER_ID=SORA_QUIZ_MASTER_ID
       - SORA_QUIZ_ATTENDEE_ID=SORA_QUIZ_ATTENDEE_ID
       - MORALIS_API_URL=http://localhost:19527
+      - MORALIS_SOLANA_API_URL=http://localhost:19527/moralis_solana/
```
Comment (BillyWooo) [inline bitacross-worker/docker/docker-compose.yml:133]: Can we just use `MORALIS_API_URL` ? Add the extra part with code?

File: tee-worker/litentry/core/common/src/abort_strategy.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+		let test_array = vec!["item1", "item2", "item3", "item4"];
+		let mut result = "";
+
+		let loop_result = loop_with_abort_strategy(
+			test_array,
+			|item| {
+				result = *item;
+
+				if *item == "item4" {
+					return Ok(false)
```
Comment (jonalvarezz) [inline tee-worker/litentry/core/common/src/abort_strategy.rs]: nitpick: should we pass an optional named struct for better readability? i.e. `return Ok(LoopControls.Break)`

File: tee-worker/litentry/core/data-providers/src/lib.rs [hunk 3]
```diff
@@ -308,6 +310,9 @@ impl DataProviderConfig {
 			if let Ok(v) = env::var("MORALIS_API_URL") {
 				config.set_moralis_api_url(v)?;
 			}
+			if let Ok(v) = env::var("MORALIS_SOLANA_API_URL") {
```
Comment (BillyWooo) [inline tee-worker/litentry/core/data-providers/src/lib.rs:313]: These two branches can be merged together.

**Review by BillyWooo — COMMENTED** [review]
In general, it looks very good. I left two small comments. The main idea is to introduce the less env variable the better.

**Review by Kailai-Wang — APPROVED** [review]
I only checked the `AbortStrategy` impl part, looks good!

**Review by BillyWooo — CHANGES_REQUESTED** [review]
If there is no strong reason, I still insist to `MORALIS_SOLANA_API_URL`.

File: bitacross-worker/docker/docker-compose.yml (hunk 1 above)
Comment (wli-pro) [reply bitacross-worker/docker/docker-compose.yml:133]: moralis_api_url is totally different from moralis_solana_api_url, I think this way is better than merging them together
moralis_api_url: "https://deep-index.moralis.io/api/v2.2/".to_string(),
moralis_solana_api_url: "https://solana-gateway.moralis.io/".to_string(),

File: tee-worker/litentry/core/data-providers/src/lib.rs (hunk 3 above)
Comment (wli-pro) [reply tee-worker/litentry/core/data-providers/src/lib.rs:313]: same as above, not worth merging them

File: tee-worker/litentry/core/common/src/abort_strategy.rs (hunk 2 above)
Comment (wli-pro) [reply tee-worker/litentry/core/common/src/abort_strategy.rs]: resolved now

**Review by BillyWooo — APPROVED** [review]

**Pull request merged**